### PR TITLE
Add (Core\Player): Added createplayerinfo_cast_spell support cast spell for some class spells

### PR DESCRIPTION
**Pull request by acidmanifesto**
(Warrior's Battle Stance and Death Knight's Blood Presence) on first login.

## Changes Proposed:
-  Implement TC's playercreateinfo_cast_spell system and tables
-  Added support cast spell for some class spells (Warrior's Battle Stance and Death Knight's Blood Presence) on first login.

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9439

## SOURCE:

https://github.com/TrinityCore/TrinityCore/commit/c8eb69df10b36d19a77743a037e700b001c8ed77
https://github.com/TrinityCore/TrinityCore/commit/30313fba8dbf5dbcdf4e91e1ea179ad1b8c9bdc5

## Tests Performed:

- Builds, Compiles, and Runs fine.

## How to Test the Changes:

1. Login with new created character class Death Knight or Warrior
2. Ensure DK has Blood Presence
3. Ensure Warrior has Battle Stance

## Known Issues and TODO List:

- [x] Independent Testing

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1638401383244389098');
+
+DROP TABLE IF EXISTS `playercreateinfo_cast_spell`;
+CREATE TABLE IF NOT EXISTS `playercreateinfo_cast_spell` (
+  `raceMask` INT(10) UNSIGNED NOT NULL DEFAULT '0',
+  `classMask` INT(10) UNSIGNED NOT NULL DEFAULT '0',
+  `spell` MEDIUMINT(8) UNSIGNED NOT NULL DEFAULT '0',
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1638401383244389098.sql:7]: Dont specify width

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  `note` VARCHAR(255) DEFAULT NULL
+) ENGINE=MYISAM DEFAULT CHARSET=utf8;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1638401383244389098.sql:9]: utf8mb4

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+DELETE FROM `playercreateinfo_cast_spell` WHERE `spell` IN (48266, 2457);
+INSERT INTO `playercreateinfo_cast_spell` (`racemask`, `classmask`, `spell`, `note`) VALUES
+(0, 32, 48266, 'Death Knight - Blood Presence'),
+(0, 1, 2457, 'Warrior - Battle Stance');
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1638401383244389098.sql:14]: Newline

**Review by Kitzunu — CHANGES_REQUESTED** [review]

**Comment by TheLegendaryKingdom** [comment]
Tested it and it works :
-New Warriors and DK correctly cast their Battle stance / Blood presence at first login after character creation.

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql (hunk 1 above)
Comment (acidmanifesto) [reply data/sql/updates/pending_db_world/rev_1638401383244389098.sql]: done

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql (hunk 3 above)
Comment (acidmanifesto) [reply data/sql/updates/pending_db_world/rev_1638401383244389098.sql]: done

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql (hunk 2 above)
Comment (acidmanifesto) [reply data/sql/updates/pending_db_world/rev_1638401383244389098.sql]: done

**Comment by acidmanifesto** [comment]
> Tested it and it works : -New Warriors and DK correctly cast their Battle stance / Blood presence at first login after character creation.

Thank you for testing!

**Comment by IntelligentQuantum** [comment]
https://github.com/azerothcore/azerothcore-wotlk/pull/9268

**Comment by ratkosrb** [comment]
> Added support cast spell for some class spells (Warrior's Battle Stance and Death Knight's Blood Presence) on first login.

Shouldn't that be done through the "Cast When Learned" spell attribute? Don't see why you'd need to define which spells should be auto casted on character create in a separate table, when the spells themselves have this specified in their attributes.

**Comment by TheLegendaryKingdom** [comment]
> > Added support cast spell for some class spells (Warrior's Battle Stance and Death Knight's Blood Presence) on first login.
> 
> Shouldn't that be done through the "Cast When Learned" spell attribute? Don't see why you'd need to define which spells should be auto casted on character create in a separate table, when the spells themselves have this specified in their attributes.

It actually doesn't work here, so it's not might not be reliable enough.
Anyway, it also avoids making a module or modifying the core to learn custom spells or customly selected spells in a sortable way, as AzerothCore is for accessibility towards customisation. I could be better to handle it that way (:

**Comment by TheLegendaryKingdom** [comment]
> 

This may explain why it was not cast anymore.
I hope this time it will be merged. (:

**Comment by TheLegendaryKingdom** [comment]
> #9268
> 
> Delete character_creation.cpp

This may explain why it was not auto-cast anymore. I hope this time it will be merged (:

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 4]
```diff
@@ -986,7 +986,13 @@ void WorldSession::HandlePlayerLoginFromDB(LoginQueryHolder const& holder)
     bool firstLogin = pCurrChar->HasAtLoginFlag(AT_LOGIN_FIRST);
     if (firstLogin)
     {
-        pCurrChar->RemoveAtLoginFlag(AT_LOGIN_FIRST);
+        {
```
Comment (IntelligentQuantum) [inline src/server/game/Handlers/CharacterHandler.cpp:989]: Remove bracket

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 5]
```diff
@@ -986,7 +986,13 @@ void WorldSession::HandlePlayerLoginFromDB(LoginQueryHolder const& holder)
...
+            pCurrChar->RemoveAtLoginFlag(AT_LOGIN_FIRST);
+
+            PlayerInfo const* info = sObjectMgr->GetPlayerInfo(pCurrChar->getRace(), pCurrChar->getClass());
+            for (uint32 spellId : info->castSpells)
+                pCurrChar->CastSpell(pCurrChar, spellId, true);
+        }
```
Comment (IntelligentQuantum) [inline src/server/game/Handlers/CharacterHandler.cpp:995]: Remove bracket

**Review by IntelligentQuantum — CHANGES_REQUESTED** [review]

**Comment by TheLegendaryKingdom** [comment]
> #9268
> 
> Delete character_creation.cpp

Sorry, the first tome I did not understand what you meant.
So, we need to remove the file `src/scripts/World/character_creation.cpp`
and the line 52 :
`AddSC_character_creation()`
in the file `src/scripts/World/world_script_loader.cpp`

**Comment by IntelligentQuantum** [comment]
> > #9268
> > Delete character_creation.cpp
> 
> Sorry, the first tome I did not understand what you meant. So, we need to remove the file `src/scripts/World/character_creation.cpp` and the line 52 : `AddSC_character_creation();` in the file `src/scripts/World/world_script_loader.cpp`

yes

File: src/server/game/Handlers/CharacterHandler.cpp (hunk 4 above)
Comment (acidmanifesto) [reply src/server/game/Handlers/CharacterHandler.cpp:989]: currently in the process of rebuilding my db and need to drop a few thigns off at a warehouse. when i am back i will test with the brackets and without the brackets.

Comment (acidmanifesto) [reply src/server/game/Handlers/CharacterHandler.cpp]: done

File: src/server/game/Handlers/CharacterHandler.cpp (hunk 5 above)
Comment (acidmanifesto) [reply src/server/game/Handlers/CharacterHandler.cpp]: done

**Review by IntelligentQuantum — APPROVED** [review]

**Review by Kitzunu — APPROVED** [review]

**Review by Winfidonarleyan — APPROVED** [review]

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 6]
```diff
@@ -982,8 +982,13 @@ void WorldSession::HandlePlayerLoginFromDB(LoginQueryHolder const& holder)
     bool firstLogin = pCurrChar->HasAtLoginFlag(AT_LOGIN_FIRST);
     if (firstLogin)
     {
+
         pCurrChar->RemoveAtLoginFlag(AT_LOGIN_FIRST);
 
+        PlayerInfo const* info = sObjectMgr->GetPlayerInfo(pCurrChar->getRace(), pCurrChar->getClass());
+        for (uint32 spellId : info->castSpells)
```
Comment (Nyeriah) [inline src/server/game/Handlers/CharacterHandler.cpp:989]: brackets

File: src/server/game/Globals/ObjectMgr.cpp [hunk 7]
```diff
@@ -3607,6 +3607,61 @@ void ObjectMgr::LoadPlayerInfo()
...
     }
 
+    // Load playercreate cast spell
+    LOG_INFO("server.loading", "Loading Player Create Cast Spell Data...");
+    {
+        uint32 oldMSTime = getMSTime();
+
+        QueryResult result = WorldDatabase.PQuery("SELECT raceMask, classMask, spell FROM playercreateinfo_cast_spell");
+
+        if (!result)
```
Comment (Nyeriah) [inline src/server/game/Globals/ObjectMgr.cpp:3617]: brackets

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql [hunk 8]
```diff
@@ -0,0 +1,14 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1638401383244389098');
+
+DROP TABLE IF EXISTS `playercreateinfo_cast_spell`;
+CREATE TABLE IF NOT EXISTS `playercreateinfo_cast_spell` (
+  `raceMask` INT UNSIGNED NOT NULL DEFAULT '0',
+  `classMask` INT UNSIGNED NOT NULL DEFAULT '0',
+  `spell` MEDIUMINT UNSIGNED NOT NULL DEFAULT '0',
+  `note` VARCHAR(255) DEFAULT NULL
```
Comment (Nyeriah) [inline data/sql/updates/pending_db_world/rev_1638401383244389098.sql:8]: @Kitzunu is this one okay?

**Comment by Nyeriah** [comment]
@acidmanifesto brackets need fixing

File: data/sql/updates/pending_db_world/rev_1638401383244389098.sql (hunk 8 above)
Comment (acidmanifesto) [reply data/sql/updates/pending_db_world/rev_1638401383244389098.sql:8]: Kitzunu got with me and provided me the link to the guide.
https://www.azerothcore.org/wiki/sql-standards#tables-and-columns
i ensure it was with in compliance.

File: src/server/game/Globals/ObjectMgr.cpp (hunk 7 above)
Comment (acidmanifesto) [reply src/server/game/Globals/ObjectMgr.cpp:3617]: done

File: src/server/game/Handlers/CharacterHandler.cpp (hunk 6 above)
Comment (acidmanifesto) [reply src/server/game/Handlers/CharacterHandler.cpp:989]: done

**Pull request merged**